core/base/inc/TBuffer.h

```diff
@@ -76,27 +80,53 @@ class TBuffer : public TObject {
 
    TBuffer(EMode mode);
    TBuffer(EMode mode, Int_t bufsiz);
-   TBuffer(EMode mode, Int_t bufsiz, void *buf, Bool_t adopt = kTRUE, ReAllocCharFun_t reallocfunc = 0);
+   TBuffer(EMode mode, Int_t bufsiz, void *buf, Bool_t adopt = kTRUE, ReAllocStateFun_t reallocfunc = 0, void *reallocData = nullptr);
    virtual ~TBuffer();
 
    Int_t    GetBufferVersion() const { return fVersion; }
    Bool_t   IsReading() const { return (fMode & kWrite) == 0; }
    Bool_t   IsWriting() const { return (fMode & kWrite) != 0; }
    void     SetReadMode();
    void     SetWriteMode();
-   void     SetBuffer(void *buf, UInt_t bufsiz = 0, Bool_t adopt = kTRUE, ReAllocCharFun_t reallocfunc = 0);
-   ReAllocCharFun_t GetReAllocFunc() const;
-   void     SetReAllocFunc(ReAllocCharFun_t reallocfunc = 0);
+   void     SetBuffer(void *buf, UInt_t bufsiz = 0, Bool_t adopt = kTRUE, ReAllocStateFun_t reallocfunc = 0, void *reallocData = nullptr);
+   ReAllocStateFun_t GetReAllocFunc() const;
+   void    *GetReAllocData() const;
+   void     SetReAllocFunc(ReAllocStateFun_t reallocfunc = 0, void *reallocData = nullptr);
    void     SetBufferOffset(Int_t offset = 0) { fBufCur = fBuffer+offset; }
    void     SetParent(TObject *parent);
    TObject *GetParent()  const;
    char    *Buffer()     const { return fBuffer; }
+   char    *GetCurrent() const { return fBufCur; }
    Int_t    BufferSize() const { return fBufSize; }
    void     DetachBuffer() { fBuffer = 0; }
    Int_t    Length()     const { return (Int_t)(fBufCur - fBuffer); }
    void     Expand(Int_t newsize, Bool_t copy = kTRUE);  // expand buffer to newsize
    void     AutoExpand(Int_t size_needed);  // expand buffer to newsize
 
+////////////////////////////////////////////////////////////////////////////////
+/// Share the underlying memory allocation with another buffer.
+//
+// This causes the passed TBuffer object to share our memory buffer.  This is
+// useful if two objects want to have their own view of the TBuffer state but
+// see identical data.
+//
+// Internally, not only do both buffers get the same memory location but a
+// resize done by the "slave" buffer updates the "owner" buffer (the opposite
+// is not true!).
+//
+// This is most useful if the "slave" does all the writings and the "owner"
+// only does reads.
+//
+   virtual void       SetSlaveBuffer(TBuffer &other) { // Share the underlying memory allocation with another TBuffer object.
+       if (other.Buffer() == Buffer()) {return;}
+       Int_t other_buf_size = BufferSize();
```

Comment:
looks backwards.

Reply:
This was a source of what I assumed was a "gotcha," but depending on your intentions, it might be a bug: I wanted to use a separate `TBufferFile` to read each basket in a branch with `GetEntriesSerialized`. (This is when I was writing an iterator interface; I used a `std::vector<TBufferFile*>` to hold all "active" basket data, baskets with any of the events that I was providing to the user at a given time. Since cluster boundaries don't always align, it wouldn't always be exactly one. `TBufferFiles` that were done got cleared and moved to the top of the vector for reuse.)

As it turns out, whenever I filled the first basket from a branch with one `TBufferFile` and the second basket from the same branch with another `TBufferFile`, it would actually overwrite the first one. I traced this behavior to `SetSlaveBuffer` and concluded that different `TBufferFiles` couldn't be used on the same branch by design. If this wasn't the design, it's a bug. (I've since found other ways to do the cluster-leveling, though they involve copying whereas this method wouldn't.)